whole-tale/girder_wholetale
# BagIt exporter

This PR refactors `taleExport` to enable exporting using multiple different formats:
1. Native (see example [here](https://github.com/whole-tale/tale_serialization_formats/tree/master/examples/zip))
2. BagIt (see example [here](https://github.com/whole-tale/tale_serialization_formats/tree/master/examples/bagit/tale_id))

### Implementation notes
Every BagIt python implementation we've encountered assumes that data already resides in a single folder on disk and operates on it to create a bag. Since, it's highly ineffective approach that requires shuffling potentially large data around, we added a basic implementation that allows to create a BagIt zipfile from data in residing in an arbitrary location (in our case Girder's assetstore) and computes relevant checksums on the fly.

### How to test?
1. Deploy locally.
1. Create a Water Tale following the [quickstart](https://wholetale.readthedocs.io/en/latest/users_guide/quickstart.html). You may need to add dependencies to Tale's workspace via `apt.txt`, `postBuild` etc. (see https://github.com/craig-willis/bdbag-water-tale/tree/master/data/workspace)
1. Export the Tale as BagIt using `GET /tale/{id}/export?taleFormat=bagit`
1. Unzip the zipball and follow the instruction to run the Tale locally (basically cd {taleId} && sh ./run.sh)

NOTE: running locally performs validation using *bdbag*, but you can experiment with it yourself by running e.g.:
* `bdbag --validate=full .`
* `bdbag --validate-profile {taleId}.zip`

### TODO

- [x] compute checksums for metadata (`tagmanifest`)
- [x] generate `bag-info.txt`
- [x] generate `fetch.txt`
